src/mysqlsh_helpers.py

```diff
@@ -35,84 +61,143 @@ class MySQL:
+
         Args:
-            root_password: password for the 'root' user
-            cluster_admin_user: user name for the cluster admin user
             cluster_admin_password: password for the cluster admin user
+            cluster_admin_user: user name for the cluster admin user
+            cluster_name: cluster name
             instance_address: address of the targeted instance
+            root_password: password for the 'root' user
+            server_config_password: password for the server config user
+            server_config_user: user name for the server config user
         """
-        self.root_password = root_password
-        self.cluster_admin_user = cluster_admin_user
         self.cluster_admin_password = cluster_admin_password
+        self.cluster_admin_user = cluster_admin_user
+        self.cluster_name = cluster_name
         self.instance_address = instance_address
+        self.root_password = root_password
+        self.server_config_password = server_config_password
+        self.server_config_user = server_config_user
+
+        try:
+            self._ensure_mysqlsh_common_dir()
+        except subprocess.CalledProcessError as e:
+            logger.exception(
+                f"Failed to ensure mysqlsh common dir for: {self.instance_address}", exc_info=e
+            )
+            raise MySQLInitializationError(e.stderr)
 
     @property
-    def mysqlsh_bin(self) -> str:
+    def _mysqlsh_bin(self) -> str:
         """Determine binary path for MySQL Shell.
 
         Returns:
             Path to binary mysqlsh
         """
         # Allow for various versions of the mysql-shell snap
         # When we get the alias use /snap/bin/mysqlsh
-        _paths = ("/usr/bin/mysqlsh", "/snap/bin/mysqlsh", "/snap/bin/mysql-shell.mysqlsh")
+        paths = ("/usr/bin/mysqlsh", "/snap/bin/mysqlsh", "/snap/bin/mysql-shell.mysqlsh")
 
-        for path in _paths:
+        for path in paths:
             if os.path.exists(path):
                 return path
+
         # Default to the full path version
         return "/snap/bin/mysql-shell"
 
     @property
-    def mysqlsh_common_dir(self) -> str:
+    def _mysqlsh_common_dir(self) -> str:
         """Determine snap common dir for mysqlsh.
 
+        Raises MySQLUpdateConfigurationError if there was an issue configuring the mysql service.
+
         Returns:
             Path to common dir
         """
-        if os.path.exists("/root/snap/mysql-shell/common"):
-            return "/root/snap/mysql-shell/common"
-        return "/tmp"
+        return "/root/snap/mysql-shell/common"
+
+    def update_mysql_configuration(self):
+        """Add a configuration file for mysqld and restart the mysql service."""
+        try:
+            # target file starts with 'z-' so it has priority over the default config file
+            shutil.copyfile("templates/mysqld.cnf", "/etc/mysql/mysql.conf.d/z-custom-mysqld.cnf")
```

Comment:
why not put the file before starting installing the mysqld?
some options change the binary format and can be set only before the creation of MySQL dir

Reply:
we now copy the config file before installing mysql-server in [37db8b5](https://github.com/canonical/mysql-operator/pull/6/commits/37db8b57689305a169c643d92613f15416ef8710)